packages/contrast-colors/index.js

```diff
@@ -13,7 +13,480 @@ governing permissions and limitations under the License.
+          let rgbArray = [d3.rgb(color.colorScale[d]).r, d3.rgb(color.colorScale[d]).g, d3.rgb(color.colorScale[d]).b];
+          let ca = contrast(rgbArray, bgRgbArray, baseV).toFixed(2);
+      
+          return Number(ca);
+        });
+      
+        contrasts = contrasts.filter(el => el != null);
+
+        let contrastColors = [];
+        let ratioLength;
+        let ratioValues;
+
+        if(Array.isArray(color.ratios)) {
+          ratioLength = color.ratios.length;
+          ratioValues = color.ratios;
+        } else if (!Array.isArray(color.ratios)){
+          ratioLength = Object.keys(color.ratios).length;
+          swatchNames = Object.keys(color.ratios);
+          ratioValues = Object.values(color.ratios);
+        }
+
+        // modify target ratio based on contrast multiplier
+        let newRatioValues = ratioValues.map(ratio => multiplyRatios(ratio, this._contrast) );
+        if(this._contrast !==1) ratioValues = newRatioValues;
+
+        // Return color matching target ratio, or closest number
+        for (let i=0; i < ratioLength; i++){
+          // Find the index of each target ratio in the array of all possible contrasts
+          let r = getMatchingRatioIndex(contrasts, ratioValues[i]);
+          let match = color.colorScale[r];
+
+          // Use the index from matching contrasts (r) to index the corresponding
+          // color value from the color scale array.
+          // use convertColorValue function to convert each color to the specified 
+          // output format and push to the new array 'contrastColors'
+          contrastColors.push(convertColorValue(match, this._output));
+        }
+
+        for (let i=0; i < contrastColors.length; i++) {
+          let n;
+          if(!swatchNames) {
+            let rVal = ratioName(color.ratios)[i];
+            n = color.name.concat(rVal);
+          }
+          else {
+            n = swatchNames[i];
+          }
+  
+          let obj = {
+            name: n,
+            contrast: ratioValues[i],
+            value: contrastColors[i]
+          };
+          newArr.push(obj);
+          // Push the same value to the returnColorValues array
+          returnColorValues.push(contrastColors[i]);
+
+        }
+        returnColors.push(colorObj);
+      }
+    });
+    this._contrastColorValues = returnColorValues;
+    this._contrastColors = returnColors;
+    return this._contrastColors;
+  }
+
+  _findContrastColorValues() {
+    return this._contrastColorValues;
+  }
+}
+
+class Color {
+  constructor({name, colorKeys, colorspace = 'RGB', ratios, smooth = false, output = 'HEX'}) { 
+    this._name = name;
+    this._colorKeys = colorKeys;
+    this._colorspace = colorspace;
+    this._ratios = ratios;
+    this._smooth = smooth;
+    this._output = output;
+    if (!this._name) {
+      throw new Error('Color missing name');
+    } 
+    if (!this._colorKeys) {
+      throw new Error(`Color Keys are undefined`);
+    }
```

Comment:
Note that `if (0)` evaluates to false, so it's possible that someone has passed a defined value here, but your error message says it's undefined. You may want to check:

1. If it's actually undefined
```js
if (this._colorKeys === undefined)
```

2. If it's an array as you expect it to be:
```js
if (Array.isArray(this._colorKeys)
```